language/stdlib/modules/LibraAccount.move

```diff
@@ -945,6 +935,43 @@ module LibraAccount {
         }
     }
 
+    // The success_epilogue is invoked at the end of successfully executed transactions.
+    fun success_epilogue<Token>(
+        account: &signer,
+        txn_sequence_number: u64,
+        txn_gas_price: u64,
+        txn_max_gas_units: u64,
+        gas_units_remaining: u64
+    ) acquires T, Balance, AccountOperationsCapability {
+        let sender = Signer::address_of(account);
+        // Load the transaction sender's account and balance resources
+        let sender_balance = borrow_global_mut<Balance<Token>>(sender);
+
+        // Charge for gas
+        let transaction_fee_amount = txn_gas_price * (txn_max_gas_units - gas_units_remaining);
+        Transaction::assert(
+            balance_for(sender_balance) >= transaction_fee_amount,
+            6
+        );
+        epilogue<Token>(sender, transaction_fee_amount, txn_sequence_number);
+    }
+
+    // The failure_epilogue is invoked at the end of transactions when the transaction is aborted during execution or
+    // during `success_epilogue`.
+    fun failure_epilogue<Token>(
+        account: &signer,
+        txn_sequence_number: u64,
+        txn_gas_price: u64,
+        txn_max_gas_units: u64,
+        gas_units_remaining: u64
+    ) acquires T, Balance, AccountOperationsCapability {
+        let sender = Signer::address_of(account);
+        // Charge for gas
+        let transaction_fee_amount = txn_gas_price * (txn_max_gas_units - gas_units_remaining);
+
+        epilogue<Token>(sender, transaction_fee_amount, txn_sequence_number);
```

Comment:
success_epilogue and failure_epilogue main difference are not asserting 'balance >= transaction_fee_amount' ? what happened if user balance is < transaction_fee_amount ? abort at withdraw_from_balance ?

Reply:
That's a great question! The failure epilogue will be executed from clean state. So the assertion should never fail as we've already check the balance at prologue and nothing is done afterwards.